pkg/store/limiter.go

```diff
@@ -82,16 +86,94 @@ func NewChunksLimiterFactory(limit uint64) ChunksLimiterFactory {
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewSeriesLimiterFactory makes a new SeriesLimiterFactory with a static limit.
 func NewSeriesLimiterFactory(limit uint64) SeriesLimiterFactory {
 	return func(failedCounter prometheus.Counter) SeriesLimiter {
 		return NewLimiter(limit, failedCounter)
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewBytesLimiterFactory makes a new BytesLimiterFactory with a static limit.
 func NewBytesLimiterFactory(limit units.Base2Bytes) BytesLimiterFactory {
 	return func(failedCounter prometheus.Counter) BytesLimiter {
 		return NewLimiter(uint64(limit), failedCounter)
 	}
 }
+
+// SeriesSelectLimits are limits applied against individual Series calls.
+type SeriesSelectLimits struct {
+	SeriesPerRequest  uint64
+	SamplesPerRequest uint64
+}
+
+func (l *SeriesSelectLimits) RegisterFlags(cmd extkingpin.FlagClause) {
+	cmd.Flag("store.grpc.series-limit", "The maximum series allowed for a single Series request. The Series call fails if this limit is exceeded. 0 means no limit.").Default("0").Uint64Var(&l.SeriesPerRequest)
+	cmd.Flag("store.grpc.samples-limit", "The maximum samples allowed for a single Series request, The Series call fails if this limit is exceeded. 0 means no limit. NOTE: For efficiency the limit is internally implemented as 'chunks limit' considering each chunk contains a maximum of 120 samples.").Default("0").Uint64Var(&l.SamplesPerRequest)
+}
+
+var _ storepb.StoreServer = &limitedStoreServer{}
+
+// limitedStoreServer is a storepb.StoreServer that can apply series and sample limits against individual Series requests.
+type limitedStoreServer struct {
+	storepb.StoreServer
+	newSeriesLimiter      SeriesLimiterFactory
+	newSamplesLimiter     ChunksLimiterFactory
+	failedRequestsCounter *prometheus.CounterVec
+}
+
+// NewLimitedStoreServer creates a new limitedStoreServer.
+func NewLimitedStoreServer(store storepb.StoreServer, reg prometheus.Registerer, selectLimits SeriesSelectLimits) storepb.StoreServer {
+	return &limitedStoreServer{
+		StoreServer:       store,
+		newSeriesLimiter:  NewSeriesLimiterFactory(selectLimits.SeriesPerRequest),
+		newSamplesLimiter: NewChunksLimiterFactory(selectLimits.SamplesPerRequest),
+		failedRequestsCounter: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_selects_dropped_total",
+			Help: "Number of select queries that were dropped due to configured limits.",
+		}, []string{"reason"}),
```

Comment:
FYI we already have this inside inside the limiters created on the lines above. Again, I like more the one introduced here for its standard potential and I think we can deprecate the old one later.